File: app/src/main/java/com/yhjoo/dochef/activities/MainActivity.java
package com.yhjoo.dochef.activities;

import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.Gravity;
import android.view.Menu;
import android.view.MenuItem;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AlertDialog;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentPagerAdapter;

import com.google.android.gms.ads.MobileAds;
import com.google.android.material.tabs.TabLayout;
import com.google.firebase.analytics.FirebaseAnalytics;
import com.skydoves.powermenu.MenuAnimation;
import com.skydoves.powermenu.PowerMenu;
import com.skydoves.powermenu.PowerMenuItem;
import com.yhjoo.dochef.R;
import com.yhjoo.dochef.adapter.MainFragmentAdapter;
import com.yhjoo.dochef.databinding.AMainBinding;
import com.yhjoo.dochef.fragments.MainInitFragment;
import com.yhjoo.dochef.fragments.MainMyRecipeFragment;
import com.yhjoo.dochef.fragments.MainRecipesFragment;
import com.yhjoo.dochef.fragments.MainTimelineFragment;
import com.yhjoo.dochef.fragments.MainUserFragment;
import com.yhjoo.dochef.utils.Utils;

import java.util.ArrayList;

public class MainActivity extends BaseActivity {
    int[] tabIcons = new int[]{R.drawable.ic_home_white, R.drawable.ic_hot_white,
            R.drawable.ic_favorite_white, R.drawable.ic_article_white, R.drawable.ic_person_white};

    AMainBinding binding;
    FirebaseAnalytics mFirebaseAnalytics;
    MainFragmentAdapter mainFragmentAdapter;
    PowerMenu powerMenu;

    MenuItem menu_notification;
    MenuItem menu_search;
    MenuItem menu_sort;
    MenuItem menu_write_recipe;
    MenuItem menu_write_post;
    MenuItem menu_setting;

    String userID;

    /*
        TODO
     */

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = AMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        setSupportActionBar(binding.mainToolbar);
        ActionBar actionBar = getSupportActionBar();
        actionBar.setDisplayHomeAsUpEnabled(false);
        actionBar.setDisplayShowTitleEnabled(false);

        MobileAds.initialize(this);
        mFirebaseAnalytics = FirebaseAnalytics.getInstance(this);

        userID = Utils.getUserBrief(this).getUserID();

        ArrayList<Fragment> fragments = new ArrayList<>();
        fragments.add(new MainInitFragment());
        fragments.add(new MainRecipesFragment());
        fragments.add(new MainMyRecipeFragment());
        fragments.add(new MainTimelineFragment());
        fragments.add(new MainUserFragment());

        mainFragmentAdapter = new MainFragmentAdapter(
                getSupportFragmentManager(),
                FragmentPagerAdapter.BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT,
                fragments);

        binding.mainViewpager.setOffscreenPageLimit(3);
        binding.mainViewpager.setPageMargin(15);
        binding.mainViewpager.addOnPageChangeListener(new TabLayout.TabLayoutOnPageChangeListener(binding.mainTablayout));
        binding.mainViewpager.setAdapter(mainFragmentAdapter);

        for (int drawable : tabIcons) {
            binding.mainTablayout.addTab(binding.mainTablayout.newTab().setIcon(drawable));
        }

        binding.mainTablayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                binding.mainViewpager.setCurrentItem(tab.getPosition());

                menu_notification.setVisible(false);
                menu_sort.setVisible(false);
                menu_write_recipe.setVisible(false);
                menu_write_post.setVisible(false);
                menu_search.setVisible(false);
                menu_setting.setVisible(false);
                switch (tab.getPosition()) {
                    case 0:
                        menu_notification.setVisible(true);
                        break;
                    case 1:
                        menu_sort.setVisible(true);
                        menu_search.setVisible(true);
                        break;
                    case 2:
                        menu_write_recipe.setVisible(true);
                        menu_search.setVisible(true);
                        break;
                    case 3:
                        menu_write_post.setVisible(true);
                        menu_search.setVisible(true);
                        break;
                    case 4:
                        menu_setting.setVisible(true);
                        break;
                }
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {
            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {
            }
        });

        powerMenu = new PowerMenu.Builder(this)
                .addItem(new PowerMenuItem("최신순", true))
                .addItem(new PowerMenuItem("인기순", false))
                .addItem(new PowerMenuItem("별점순", false))
                .setAnimation(MenuAnimation.SHOWUP_TOP_RIGHT)
                .setMenuRadius(10f)
                .setMenuShadow(0f)
                .setTextColor(ContextCompat.getColor(this, R.color.colorPrimary))
                .setSelectedTextColor(Color.WHITE)
                .setTextGravity(Gravity.CENTER)
                .setMenuColor(Color.WHITE)
                .setSelectedMenuColor(ContextCompat.getColor(this, R.color.colorPrimary))
                .setBackgroundAlpha(0f)
                .build();

        powerMenu.setOnMenuItemClickListener((position, item) -> {
            if (position == 0)
                sortMenu(MainRecipesFragment.SORT.LATEST);
            else if (position == 1)
                sortMenu(MainRecipesFragment.SORT.POPULAR);
            else if (position == 2)
                sortMenu(MainRecipesFragment.SORT.RATING);
            powerMenu.setSelectedPosition(position);
            powerMenu.dismiss();
        });
    }

    @Override
    public void onBackPressed() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);

        builder.setMessage("종료하시겠습니까?")
                .setPositiveButton("종료", (dialog, which) -> {
                    Bundle bundle = new Bundle();
                    bundle.putString(FirebaseAnalytics.Param.ITEM_ID, getString(R.string.analytics_id_terminated));
                    bundle.putString(FirebaseAnalytics.Param.ITEM_NAME, getString(R.string.analytics_name_terminated));
                    bundle.putString(FirebaseAnalytics.Param.CONTENT_TYPE, getString(R.string.analytics_type_text));
                    mFirebaseAnalytics.logEvent(getString(R.string.analytics_event_terminated), bundle);

                    dialog.dismiss();
                    finish();
                })
                .setNegativeButton("취소", (dialog, which) -> dialog.dismiss())
                .show();
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main, menu);

        menu_notification = menu.findItem(R.id.main_menu_notification);
        menu_sort = menu.findItem(R.id.main_menu_sort);
        menu_write_recipe = menu.findItem(R.id.main_menu_write_recipe);
        menu_write_post = menu.findItem(R.id.main_menu_write_post);
        menu_search = menu.findItem(R.id.main_menu_search);
        menu_setting = menu.findItem(R.id.main_menu_setting);

        menu_sort.setVisible(false);
        menu_write_recipe.setVisible(false);
        menu_write_post.setVisible(false);
        menu_search.setVisible(false);
        menu_setting.setVisible(false);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.main_menu_notification:
                startActivity(new Intent(this, NotificationActivity.class));
                break;
            case R.id.main_menu_sort:
                powerMenu.showAsAnchorRightBottom(binding.mainToolbar);
                break;
            case R.id.main_menu_write_recipe:
                startActivity(new Intent(this, RecipeMakeActivity.class));
                break;
            case R.id.main_menu_write_post:
                Intent intent = new Intent(this, PostWriteActivity.class)
                        .putExtra("MODE", PostWriteActivity.MODE.WRITE);
                startActivity(intent);
                break;
            case R.id.main_menu_search:
                startActivity(new Intent(this, SearchActivity.class));
                break;
            case R.id.main_menu_setting:
                startActivity(new Intent(this, SettingActivity.class));
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    void sortMenu(MainRecipesFragment.SORT sort) {
        ((MainRecipesFragment) mainFragmentAdapter.getItem(1)).changeSortMode(sort);
    }
}

File: app/src/main/java/com/yhjoo/dochef/ui/activities/MainActivity.kt
package com.yhjoo.dochef.ui.activities

import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.view.Gravity
import android.view.Menu
import android.view.MenuItem
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentPagerAdapter
import com.afollestad.materialdialogs.MaterialDialog
import com.google.android.gms.ads.MobileAds
import com.google.android.material.tabs.TabLayout
import com.google.firebase.analytics.FirebaseAnalytics
import com.skydoves.powermenu.MenuAnimation
import com.skydoves.powermenu.OnMenuItemClickListener
import com.skydoves.powermenu.PowerMenu
import com.skydoves.powermenu.PowerMenuItem
import com.yhjoo.dochef.R
import com.yhjoo.dochef.activities.*
import com.yhjoo.dochef.ui.adapter.MainFragmentAdapter
import com.yhjoo.dochef.databinding.AMainBinding
import com.yhjoo.dochef.ui.fragments.*
import com.yhjoo.dochef.utils.Utils
import java.util.*

class MainActivity : BaseActivity() {
    private val tabIcons = intArrayOf(
        R.drawable.ic_home_white, R.drawable.ic_hot_white,
        R.drawable.ic_favorite_white, R.drawable.ic_article_white, R.drawable.ic_person_white
    )

    val binding: AMainBinding by lazy { AMainBinding.inflate(layoutInflater) }
    private lateinit var mFirebaseAnalytics: FirebaseAnalytics
    private lateinit var mainFragmentAdapter: MainFragmentAdapter

    private lateinit var powerMenu: PowerMenu
    lateinit var menuNotification: MenuItem
    lateinit var menuSearch: MenuItem
    lateinit var menuSort: MenuItem
    lateinit var menuWriteRecipe: MenuItem
    lateinit var menuWritePost: MenuItem
    lateinit var menuSetting: MenuItem
    lateinit var userID: String

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)
        setSupportActionBar(binding.mainToolbar)

        val actionBar = supportActionBar!!
        actionBar.setDisplayHomeAsUpEnabled(false)
        actionBar.setDisplayShowTitleEnabled(false)

        MobileAds.initialize(this)
        mFirebaseAnalytics = FirebaseAnalytics.getInstance(this)

        userID = Utils.getUserBrief(this).userID
        val fragments = ArrayList<Fragment>().apply {
            add(MainInitFragment())
            add(MainRecipesFragment())
            add(MainMyRecipeFragment())
            add(MainTimelineFragment())
            add(MainUserFragment())
        }
        mainFragmentAdapter = MainFragmentAdapter(
            supportFragmentManager,
            FragmentPagerAdapter.BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT,
            fragments
        )

        binding.apply {
            mainViewpager.offscreenPageLimit = 3
            mainViewpager.pageMargin = 15
            mainViewpager.addOnPageChangeListener(
                TabLayout.TabLayoutOnPageChangeListener(
                    mainTablayout
                )
            )
            mainViewpager.adapter = mainFragmentAdapter
            for (drawable in tabIcons) {
                mainTablayout.addTab(mainTablayout.newTab().setIcon(drawable))
            }
            mainTablayout.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
                override fun onTabSelected(tab: TabLayout.Tab) {
                    mainViewpager.currentItem = tab.position
                    menuNotification.isVisible = false
                    menuSort.isVisible = false
                    menuWriteRecipe.isVisible = false
                    menuWritePost.isVisible = false
                    menuSearch.isVisible = false
                    menuSetting.isVisible = false
                    when (tab.position) {
                        0 -> menuNotification.isVisible = true
                        1 -> {
                            menuSort.isVisible = true
                            menuSearch.isVisible = true
                        }
                        2 -> {
                            menuWriteRecipe.isVisible = true
                            menuSearch.isVisible = true
                        }
                        3 -> {
                            menuWritePost.isVisible = true
                            menuSearch.isVisible = true
                        }
                        4 -> menuSetting.isVisible = true
                    }
                }

                override fun onTabUnselected(tab: TabLayout.Tab) {}
                override fun onTabReselected(tab: TabLayout.Tab) {}
            })
        }

        powerMenu = PowerMenu.Builder(this)
            .addItem(PowerMenuItem("최신순", true))
            .addItem(PowerMenuItem("인기순", false))
            .addItem(PowerMenuItem("별점순", false))
            .setAnimation(MenuAnimation.SHOWUP_TOP_RIGHT)
            .setMenuRadius(10f)
            .setMenuShadow(0f)
            .setTextColor(ContextCompat.getColor(this, R.color.colorPrimary))
            .setSelectedTextColor(Color.WHITE)
            .setTextGravity(Gravity.CENTER)
            .setMenuColor(Color.WHITE)
            .setSelectedMenuColor(ContextCompat.getColor(this, R.color.colorPrimary))
            .setBackgroundAlpha(0f)
            .build()
        powerMenu.onMenuItemClickListener =
            OnMenuItemClickListener { position: Int, _: PowerMenuItem? ->
                when (position) {
                    0 -> sortMenu(MainRecipesFragment.VALUES.SORT.LATEST)
                    1 -> sortMenu(MainRecipesFragment.VALUES.SORT.POPULAR)
                    2 -> sortMenu(MainRecipesFragment.VALUES.SORT.RATING)
                }
                powerMenu.selectedPosition = position
                powerMenu.dismiss()
            }
    }

    override fun onBackPressed() {
        MaterialDialog(this).show {
            message(text = "종료하시겠습니까?")
            positiveButton(text = "확인") {
                val bundle = Bundle()
                bundle.apply {
                    putString(
                        FirebaseAnalytics.Param.ITEM_ID,
                        getString(R.string.analytics_id_terminated)
                    )
                    putString(
                        FirebaseAnalytics.Param.ITEM_NAME,
                        getString(R.string.analytics_name_terminated)
                    )
                    putString(
                        FirebaseAnalytics.Param.CONTENT_TYPE,
                        getString(R.string.analytics_type_text)
                    )
                }

                mFirebaseAnalytics.logEvent(
                    getString(R.string.analytics_event_terminated),
                    bundle
                )
                finish()
            }
            negativeButton(text = "취소")
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main, menu)

        menuNotification = menu.findItem(R.id.main_menu_notification)
        menuSort = menu.findItem(R.id.main_menu_sort)
        menuWriteRecipe = menu.findItem(R.id.main_menu_write_recipe)
        menuWritePost = menu.findItem(R.id.main_menu_write_post)
        menuSearch = menu.findItem(R.id.main_menu_search)
        menuSetting = menu.findItem(R.id.main_menu_setting)

        menuSort.isVisible = false
        menuWriteRecipe.isVisible = false
        menuWritePost.isVisible = false
        menuSearch.isVisible = false
        menuSetting.isVisible = false

        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.main_menu_notification -> startActivity(
                Intent(
                    this,
                    NotificationActivity::class.java
                )
            )
            R.id.main_menu_sort -> powerMenu.showAsAnchorRightBottom(binding.mainToolbar)
            R.id.main_menu_write_recipe -> startActivity(
                Intent(
                    this,
                    RecipeMakeActivity::class.java
                )
            )
            R.id.main_menu_write_post -> {
                val intent = Intent(this, PostWriteActivity::class.java)
                    .putExtra("MODE", PostWriteActivity.VALUES.UIMODE.WRITE)
                startActivity(intent)
            }
            R.id.main_menu_search -> startActivity(Intent(this, SearchActivity::class.java))
            R.id.main_menu_setting -> startActivity(Intent(this, SettingActivity::class.java))
        }
        return super.onOptionsItemSelected(item)
    }

    private fun sortMenu(sort: Int) {
        (mainFragmentAdapter.getItem(1) as MainRecipesFragment).changeSortMode(sort)
    }
}